services/shhext/chat/protocol.go

```diff
@@ -39,50 +57,110 @@ func (p *ProtocolService) addBundleAndMarshal(myIdentityKey *ecdsa.PrivateKey, m
 		// an issue anymore
 		msg.Bundle = bundle
 	} else {
-		msg.Bundles = []*Bundle{bundle}
-	}
-
-	// marshal for sending to wire
-	marshaledMessage, err := proto.Marshal(msg)
-	if err != nil {
-		p.log.Error("encryption-service", "error marshaling message", err)
-		return nil, err
+		msg.Bundles = []*protobuf.Bundle{bundle}
 	}
 
-	return marshaledMessage, nil
+	return msg, nil
 }
 
 // BuildPublicMessage marshals a public chat message given the user identity private key and a payload
-func (p *ProtocolService) BuildPublicMessage(myIdentityKey *ecdsa.PrivateKey, payload []byte) ([]byte, error) {
+func (p *ProtocolService) BuildPublicMessage(myIdentityKey *ecdsa.PrivateKey, payload []byte) (*protobuf.ProtocolMessage, error) {
 	// Build message not encrypted
-	protocolMessage := &ProtocolMessage{
+	protocolMessage := &protobuf.ProtocolMessage{
 		InstallationId: p.encryption.config.InstallationID,
 		PublicMessage:  payload,
 	}
 
-	return p.addBundleAndMarshal(myIdentityKey, protocolMessage, false)
+	return p.addBundle(myIdentityKey, protocolMessage, false)
 }
 
-// BuildDirectMessage marshals a 1:1 chat message given the user identity private key, the recipient's public key, and a payload
-func (p *ProtocolService) BuildDirectMessage(myIdentityKey *ecdsa.PrivateKey, publicKey *ecdsa.PublicKey, payload []byte) ([]byte, error) {
+type ProtocolMessageSpec struct {
+	Message *protobuf.ProtocolMessage
+	// Installations is the targeted devices
+	Installations []*multidevice.Installation
+	// SharedSecret is a shared secret established among the installations
+	SharedSecret []byte
+}
+
+func (p *ProtocolMessageSpec) MinVersion() uint32 {
+
+	if len(p.Installations) == 0 {
+		return 0
+	}
+
+	version := p.Installations[0].Version
+
+	for _, installation := range p.Installations[1:] {
+		if installation.Version < version {
+			version = installation.Version
+		}
+	}
+	return version
+}
+
+func (p *ProtocolMessageSpec) PartitionedTopic() bool {
```

Comment:
to futureproof I would do an enum here instead of a boolean. `PARTITION_TOPIC_NO_SUPPORT`, `PARTITION_TOPIC_V1`, etc.